tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/Summary.java

```diff
@@ -48,57 +64,163 @@
+                && !UserRegistry.userIsAdmin(cs.user)) {
+                return Response.status(Status.FORBIDDEN).build();
             }
-            session.userDidAction();
-            // End boilerplate
-
-            VettingViewerQueue.Status status[] = new VettingViewerQueue.Status[1];
-            StringBuilder sb = new StringBuilder();
-            JSONObject jStatus = new JSONObject();
-            VettingViewerQueue vvq = VettingViewerQueue.getInstance();
-            String str = vvq.getPriorityItemsSummaryOutput(session, status, request.loadingPolicy, sb, jStatus);
-            SummaryResponse resp = new SummaryResponse();
-            resp.jStatus = jStatus;
-            resp.status = status[0];
-            resp.ret = str;
-            resp.output = sb.toString();
-            return Response.ok(resp, MediaType.APPLICATION_JSON).build();
+            return getPriorityItemsSummary(cs, request);
         } catch (JSONException | IOException ioe) {
             return Response.status(500, "An exception occurred").entity(ioe).build();
         }
     }
 
+    /**
+     * Get the response for Priority Items Summary (possibly a snapshot)
+     *
+     * Each request specifies a loading policy: START, NOSTART, or FORCESTOP.
+     * Typically the client makes a request with START, then makes repeated requests with NOSTART
+     * while the responses have status PROCESSING (or WAITING), until there is a response with status READY.
+     * The response with status READY contains the actual requested Priority Items Summary data.
+     *
+     * Each request specifies a snapshot policy: SNAP_NONE, SNAP_CREATE, or SNAP_SHOW.
+     *
+     * @param cs the CookieSession identifying the user
+     * @param request the SummaryRequest
+     * @return the Response
+     *
+     * @throws IOException
+     * @throws JSONException
+     */
+    private Response getPriorityItemsSummary(CookieSession cs, SummaryRequest request) throws IOException, JSONException {
+        cs.userDidAction();
+        if (SurveySnapshot.SNAP_SHOW.equals(request.snapshotPolicy)) {
+            return showSnapshot(request.snapshotId);
+        }
+        SummaryResponse sr = getPrItSu(cs, request.loadingPolicy);
+        if (SurveySnapshot.SNAP_CREATE.equals(request.snapshotPolicy)
+            && sr.status == VettingViewerQueue.Status.READY) {
+            saveSnapshot(sr);
+        }
+        return Response.ok(sr, MediaType.APPLICATION_JSON).build();
+    }
+
+    /**
+     * Get the response for Priority Items Summary (new, not an already existing snapshot)
+     *
+     * @param cs the CookieSession identifying the user
+     * @param loadingPolicy the LoadingPolicy
+     * @return the SummaryResponse
+     *
+     * @throws IOException
+     * @throws JSONException
+     */
+    private SummaryResponse getPrItSu(CookieSession cs, LoadingPolicy loadingPolicy) throws IOException, JSONException {
```

Comment:
minor for later: name is a little cryptic

Reply:
could probably just be getSummaryResponse()

Reply:
anyway, all getPriorityItemsSummaryOutput() really needs is the *org*. It would be better to just pass that down, not the whole cookiesession.

Then this would be getSummaryOutputForOrg()

Reply:
"not the whole cookiesession" -- actually it does need the CookieSession as a parameter for VettingViewerQueue.getPriorityItemsSummaryOutput which in turn needs it for VettingViewerQueue.getEntry -- maybe they could all be refactored away but  that's sort of tangential to this ticket -- I'm not trying (yet) to thoroughly refactor VettingViewerQueue!